projects/laji/src/app/shared/pipe/label.pipe.ts

```diff
@@ -18,100 +18,70 @@ type labelType = 'qname'|'fullUri'|'warehouse'|'withKey'|'emptyWhenMissing';
   name: 'label',
   pure: false
 })
-export class LabelPipe implements PipeTransform, OnDestroy {
-  value = '';
+export class LabelPipe implements PipeTransform {
+  value: string|string[] = '';
   lastKey: string;
-  onLangChange: Subscription;
 
   constructor(private translate: TranslateService,
               private warehouseService: WarehouseValueMappingService,
               private triplestoreLabelService: TriplestoreLabelService,
-              private _ref: ChangeDetectorRef) {
+              private cdr: ChangeDetectorRef) {
   }
 
-  updateValue(key: string, type?: labelType): void {
-    if (type === 'warehouse') {
-      this.warehouseService.getOriginalKey(key).subscribe(
-          (res: string) => {
-            if (res) {
-              this._updateValue(res);
-            } else {
-              this.value = key;
-              this._ref.markForCheck();
-            }
-          },
-          () => this._updateValue(key)
-        );
-    } else {
-      this._updateValue(key, type);
-    }
-  }
-
-  transform(value: any, type?: labelType, key?: string): any {
-    if (Array.isArray(value)) {
-      return value.map(v => this.transform(v, type, key));
-    }
-    if (value && key) {
-      value = value[key];
-    }
-    if (!value || typeof value !== 'string' || value.length === 0 ||
-       (type === 'fullUri' && value.indexOf('http') !== 0)) {
+  transform<T extends string|string[]>(value: T, type?: labelType): T {
+    if (!value || (typeof value !== 'string' && !Array.isArray(value)) || value.length === 0) {
```

Comment:
Is this check necessary when `T` is already typed to `string|string[]`? The check could be simply:

```if (!value.length) {```

Reply:
That allows typing to be so that if value = string the output will be type string, and if value is string[] then the output would be value string[] so this will prevent the case where string input and output would be string[], but method overloading might make this little pit more readable :)

Reply:
Hmm I'm not following... Can you give me an example for a `value` (that is of type `string|string[]`) that behaves differently for these checks:

```
if (!value.length) {
```

```
if (!value || (typeof value !== 'string' && !Array.isArray(value)) || value.length === 0) {
```

Reply:
Within the method there is nothing that would behave differently. Difference is that the method is returning type T so if you'd use it on a value that is of type string it would keep the return of this function as a string and not return type string|string[] for the code that's using it :)

More about those can be found here https://www.typescriptlang.org/docs/handbook/generics.html

Reply:
ps. yesterday just noticed that there is a handy strictTemplates option in tsconfig that I'm refactoring our code to use 😄

Reply:
Ok, thanks for the explanation! :)